tests/unit/test_charm.py

```diff
@@ -269,3 +313,249 @@ def test_given_any_config_when_on_config_changed_then_systemd_manager_configurat
+    @patch("subprocess.run")
+    def test_given_detach_ue_action_when_detach_ue_action_then_srsue_service_sets_action_result(
+        self, _, __
+    ):
+        mock_event = Mock()
+        mock_event.params = self.DETACH_ACTION_PARAMS
+
+        self.harness.charm._on_detach_ue_action(mock_event)
+
+        self.assertEqual(
+            mock_event.set_results.call_args,
+            call({"status": "ok", "message": "Detached successfully"}),
+        )
+
+    @patch("subprocess.run")
+    def test_given_on_remove_default_gw_action_when_default_gw_action_then_removes_default_gw(
+        self, patch_subprocess_run
+    ):
+        mock_event = Mock()
+
+        self.harness.charm._on_remove_default_gw_action(mock_event)
+
+        patch_subprocess_run.assert_any_call(
+            "route del default", shell=True, stdout=-1, encoding="utf-8"
+        )
+
+    @patch("subprocess.run")
+    def test_given_on_remove_default_gw_action_when_default_gw_action_then_sets_action_result(
+        self, patch_subprocess_run
+    ):
+        mock_event = Mock()
+
+        self.harness.charm._on_remove_default_gw_action(mock_event)
+
+        self.assertEqual(
+            mock_event.set_results.call_args,
+            call({"status": "ok", "message": "Default route removed!"}),
+        )
+
+    @patch("subprocess.run")
+    def test_given_on_remove_default_gw_action_when_remove_default_gw_action_then_status_does_not_change(  # noqa: E501
+        self, _
+    ):
+        mock_event = Mock()
+        old_status = self.harness.charm.unit.status
+
+        self.harness.charm._on_remove_default_gw_action(mock_event)
+
+        self.assertEqual(self.harness.charm.unit.status, old_status)
+
+    @patch("subprocess.run")
+    @patch("builtins.open", new_callable=mock_open)
+    def test_given_unit_in_relation_data_and_mme_is_not_ipv4_when_mme_relation_changed_then_status_does_not_change(  # noqa: E501
+        self, _, __
+    ):
+        old_status = self.harness.charm.unit.status
+        mock_event = Mock()
+        mock_event.unit = "lte-vepc/0"
+        mme_relation_id = self.harness.add_relation(relation_name="mme", remote_app="lte-vepc")
+        self.harness.add_relation_unit(mme_relation_id, "lte-vepc/0")
+        self.harness.update_relation_data(
+            relation_id=mme_relation_id,
+            app_or_unit="lte-vepc/0",
+            key_values={"mme-addr": "not an ip"},
+        )
+
+        self.assertEqual(self.harness.charm.unit.status, old_status)
+
+    @patch("subprocess.run", new=Mock())
+    @patch("builtins.open", new_callable=mock_open)
+    @patch("charm.service_active")
+    def test_given_unit_in_relation_data_and_mme_is_ipv4_when_mme_relation_changed_then_status_is_active(  # noqa: E501
+        self, patch_service_active, _
+    ):
+        valid_mme = "0.0.0.0"
+        mock_event = Mock()
+        mock_event.unit = "lte-vepc/0"
+        self.harness.charm._stored.installed = True
```

Comment:
Overall it feels weird to have to set those. Couldn't we run the actual event hook that sets those instead of setting them up manually. This way when we refactor and remove the need for _stored we won't end up having to refactor the tests. @ghislainbourgeois I'd be curious about your opinion here.

Reply:
Looking at the code, I am not even sure it makes a lot of sense to use `StoredState` for this. In the AGW charm, we validate that the install as completed by validating that the magmad service is enabled. The same could be done here with the `SRS_ENODEB_SERVICE`, and we could get rid of the `StoredState` and just mock the call or the utils function.

Reply:
There are a couple of place where the `StoredState` is used and I think it would be better to check the workload directly instead. For now, since this is mostly adding tests and doing small refactors, I think we could go with that and have another PR that gets rid of much of the `StoredState`.

Reply:
For this test I'm relying on both ways, because the `_get_current_status` method relies on both conditions `self.harness.charm._stored.installed = True` and `patch_service_active.return_value = True`, that way if we don't set both, status will be wrong. And yes, I agree with @ghislainbourgeois about checking the workload only, but not in this PR I guess.

Reply:
The problem I have with having our tests depending on stored state is that when we get rid of them, we will also have to refactor our tests. Then how de we know if we broke anything. When there is a stored state, an event triggered it. So I'd suggest triggering the events that store state instead of storing state directly.

Reply:
Looking at the code, if we decide to test through the install event, we will have to mock a bunch of real calls, and when we do refactor, we will probably have to change the tests anyway because we will be adding additional calls and potentially more mocks. This can be tried I guess, but we would have to be careful with the mocking.

Reply:
I'm trying @gruyaume 's approach, it is true that it needs a bunch of mocking, but I don't quite this part @ghislainbourgeois
> we will probably have to change the tests anyway because we will be adding additional calls and potentially more mocks

Reply:
When we replace the `StoredState` for direct checks in the workload, we will be calling additional things we will want to mock. Instead of checking for `_stored.installed`, we will be doing something like `systemctl is-enabled snsran`. This call does not exist yet and it might break the test when we do refactor.

Reply:
Got it. Yeah, I expected to have to refactor tests anyway.